extensions/ql-vscode/src/run-remote-query.ts

```diff
@@ -68,97 +72,291 @@ export async function getRepositories(): Promise<string[] | undefined> {
+  return {
+    base64Pack,
+    language
+  };
+}
+
+async function createRemoteQueriesTempDirectory() {
+  const remoteQueryDir = await tmp.dir({ dir: tmpDir.name, unsafeCleanup: true });
+  const queryPackDir = path.join(remoteQueryDir.path, 'query-pack');
+  await fs.mkdirp(queryPackDir);
+  return { remoteQueryDir, queryPackDir };
+}
+
+async function getPackedBundlePath(queryPackDir: string) {
+  return tmp.tmpName({
+    dir: path.dirname(queryPackDir),
+    postfix: 'generated.tgz',
+    prefix: 'qlpack',
+  });
+}
+
+export async function runRemoteQuery(
+  cliServer: cli.CodeQLCliServer,
+  credentials: Credentials,
+  uri: Uri | undefined,
+  dryRun: boolean,
+  progress: ProgressCallback,
+  token: CancellationToken
+): Promise<void | string> {
+  if (!(await cliServer.cliConstraints.supportsRemoteQueries())) {
+    throw new Error(`Remote queries are not supported by this version of CodeQL. Please upgrade to v${cli.CliVersionConstraint.CLI_VERSION_REMOTE_QUERIES
+      } or later.`);
   }
 
-  // Get the controller repo from the config, if it exists.
-  // If it doesn't exist, prompt the user to enter it, and save that value to the config.
-  let controllerRepo: string | undefined;
-  controllerRepo = getRemoteControllerRepo();
-  if (!controllerRepo || !REPO_REGEX.test(controllerRepo)) {
-    void logger.log(controllerRepo ? 'Invalid controller repository name.' : 'No controller repository defined.');
-    controllerRepo = await window.showInputBox({
-      title: 'Controller repository in which to display progress and results of remote queries',
-      placeHolder: '<owner>/<repo>',
-      prompt: 'Enter the name of a GitHub repository in the format <owner>/<repo>',
-      ignoreFocusOut: true,
+  const { remoteQueryDir, queryPackDir } = await createRemoteQueriesTempDirectory();
+  try {
+    if (!uri?.fsPath.endsWith('.ql')) {
+      throw new UserCancellationException('Not a CodeQL query file.');
+    }
+
+    progress({
+      maxStep: 5,
+      step: 1,
+      message: 'Determining project list'
     });
-    if (!controllerRepo) {
-      void showAndLogErrorMessage('No controller repository entered.');
-      return;
-    } else if (!REPO_REGEX.test(controllerRepo)) { // Check if user entered invalid input
-      void showAndLogErrorMessage('Invalid repository format. Must be a valid GitHub repository in the format <owner>/<repo>.');
-      return;
+
+    const queryFile = uri.fsPath;
+    const repositoriesFile = queryFile.substring(0, queryFile.length - '.ql'.length) + '.repositories';
+    let ref: string | undefined;
+    // For the case of single file remote queries, use the language from the config in order to avoid the user having to select it.
+    let fallbackLanguage: string | undefined;
+    let repositories: string[] | undefined;
+
+    progress({
+      maxStep: 5,
+      step: 2,
+      message: 'Determining query target language'
+    });
+
+    // If the user has an explicit `.repositories` file, use that.
```

Comment:
Ah, we do still have support for `.repositories`. I think that would be completely fine to remove at this point, but perhaps we should discuss in our next meeting.

Reply:
I think that's worth talking about. @shati-patel may have some opinions on this.